wip

smichr opened the pull request:
@siefkenj , I will leave some comments in the code that I have made. We can discuss those here before this gets merged to your branch, #11241.

smichr committed:
pointify modified

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -109,16 +107,50 @@ def __new__(cls, *args, **kwargs):
+                    break
+            return i + 1
+        if not args:
+            raise ValueError('most have 1 or more Points')
```
most -> must

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -906,17 +991,17 @@ class Point3D(Point):
+
+        if len(args) == 1:
+            args = args[0]
+        coords = Point.pointify(args, ensure_point=False)
```
I think you need to pass `**kwargs` to `pointify`, otherwise `eval=False` won't work.  Or am I missing something?

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -109,16 +107,50 @@ def __new__(cls, *args, **kwargs):
     is_Point = True
 
+    @staticmethod
+    def normalize_dimensions(dimension=None, *args):
```
Nice idea.  I think we could make this a little more like `pointify` if we had `dimension` as a kwarg after `*args`.  Perhaps we could even add a `pointify=False` default argument.  If `pointify=True`, `pointify` could be called on the args first.  This would save come code in most of the instances where you use `normalize_dimension`.

smichr replied:
Good idea -- done.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -109,16 +107,50 @@ def __new__(cls, *args, **kwargs):
+        >>> Point.normalize_dimensions(None, Point(1,2,0), Point(1,2,3,0))
+        [Point3D(1, 2, 0), Point3D(1, 2, 3)]
+        """
+        def size(p):
```
If I'm reading this correctly, `Point.normalize_dimension(Point(1,2,0), Point(1,3,0))` will return `[Point2D(1,2), Point2D(1,3)]` ?

siefkenj replied:
A highlevel way to do this is

```
def size(p):
    return max([1] + [i for i,x in enumerate(p) if x != 0]) + 1
```

smichr replied:
done

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -738,21 +825,19 @@ def is_concyclic(*points):
+        from .ellipse import Circle
+        c = Circle(*points[:3])
+        for point in points[3:]:
+            if point not in c:  # XXX does this raise error if it can't be determined?
```
IIRC, `in` doesn't is binary but `contains` is trinary with `True` `False` `None`

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -623,22 +716,16 @@ class Point2D(Point):
+        if len(args) == 1:
+            args = args[0]
+        coords = Point.pointify(args, dimension=2,
+            strict=True, ensure_point=False)
```
Interesting use of `ensure_point=False`

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -477,29 +580,27 @@ def intersection(self, other):
 
         """
-        other = Point.pointify(other, ensure_point=False)
+        if not isinstance(other, (Point, GeometryEntity)):
```
`Point` inherits from `GeometryEntity` so `isinstance(other, GeometryEntity)` will do. This shows up several times.

smichr replied:
go it

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -527,45 +628,37 @@ def __add__(self, other):
+        if not isinstance(other, (GeometryEntity, Point)):
+            other = Point.pointify(other,
+                dimension=self.ambient_dimension, ensure_point=False)
+        if isinstance(other, (Point, Tuple)):
```
Seeing this code makes me nervous.  Perhaps a comment saying that `pointify` will return a `GeometricEntity` or a `Tuple`?

smichr replied:
(it's gone now)

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -477,29 +580,27 @@ def intersection(self, other):
     def dot(self, p2):
         """Return dot product of self with another Point."""
-        p2 = Point.pointify(p2, dimension=self.ambient_dimension)
+        p2 = Point.pointify(p2, dimension=self.ambient_dimension, ensure_point=False)
```
Why?

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -261,59 +296,127 @@ def affine_rank(*points):
-        is carried out and the return value will be a GeometryEntity.
-        If `ensure_point` is set (default `True`), a type check for Point is carried out
-        and the return value will be a Point."""
+    def pointify(a, dimension=None, strict=False, ensure_point=True):
```
No `**kwargs` to `Point` anymore?  I think `pad=True` would be more intuitive than `strict=False`.

smichr replied:
The Point could be padded or trimmed. In the most recent pass through I have eliminated `pointify` and moved that logic into `Point`. `Point*D` are just stubs to call `Point` to do the heavy lifting.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -261,59 +296,127 @@ def affine_rank(*points):
+        Traceback (most recent call last):
+        ...
+        ValueError: wrong number of dimensions
+        """
```
Docs are great! :-)

smichr replied:
I was just following the example of whoever went through and highly upgraded the geometry docstrings some time ago. :-)

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -261,59 +296,127 @@ def affine_rank(*points):
+        if was_point:
+            a = Point(*a, evaluate=False)
+        elif ensure_point:
+            a = Point(*a)
```
This method seems to have gotten a lot more confusing...

smichr replied:
That's what I was thinking. It just didn't have the "do one thing well" design. See the updated `Point.__new__`.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -261,59 +296,127 @@ def affine_rank(*points):
+        elif not is_sequence(a):
+            raise TypeError(filldedent('''
+                expecting Point or sequence of 2 or
+                more coordinates, not `{}`'.format(a))
```
`.format()` outside of the `'''`?

smichr replied:
got it

siefkenj reviewed (commented):
I've reviewed `point.py`.  There are some nice changes, but I think we could make them nicer.  I see we've expanded the API to allow points to be padded with zeros to match the dimension of the caller in most cases.  I personally don't like this and is why I had `pointify` delegate to the `Point2D` and `Point3D` constructors to do their dirty work...I think at minimum we should raise a warning when a point is padded with zeros.  I would be all for deprecating the existing behavior with `Point3D` so in the future it has to be declared with 3 coordinates.

I also now think maybe `pointify` would be best as a class method.  As a class method, `dimension` would default to `cls.ambient_dimension`.  This would shorten a lot of calls to `pointify`.  `normalize_dimension` could also be a class method that does the same.  There'd be some subtleties in doing this since 4d and up points all share the same class, and so `ambient_dimension` cannot be a class property, but I think this can be worked around...

smichr commented:
I'll let this rest a bit. Regarding the mutation of dimensions...we already try to use the minimal object to represent a geometric entity, e.g.

``` python
>>> Triangle(*[Point(dim=2)]*3)
Point2D(0, 0)
```

That's the motivation for allowing dimensions to change. But this is definitely open to discussion and a total review to see that it's working properly. I just fixed a case where this gave problems when normalization of 3D points with trailing 0s created 2D instances which were then incompatible with the other Point that they were entering a calculation with.

smichr commented on sympy/geometry/point.py:
```diff
@@ -167,20 +224,23 @@ def is_collinear(*args):
-            return simplify(x1*y2-x2*y1) == 0
+        if s.ambient_dimension == 2:
+            (x1, y1), (x2, y2) = s.args, o.args
+            rv = (x1*y2 - x2*y1).equals(0)
```
I used `equals` here because `simplify` alone isn't powerful enough to resolve some expressions. I found some of those expression intermittently during testing and found that the use of `set` was making it not always happen. So I changed `set` to `uniq` (which preserves order) and used `permutations` in the tests to force such expressions to be encountered.

siefkenj replied:
It's unfortunate that the new `Matrix` fixes don't resolve this...I bet if we took the rank of the transpose, it would work.  If we override this for 2D, I think that should be done by overriding the method to `Point2D` rather than special-casing it in `Point`.

smichr committed:
pointify removed; Point modified; other changes

* Point now allows one to specify the desired dimension.
* Point will return Point instances unchanged except for dimension change
  (if needed)
* the ellipse docstring was modified.
* moved Undecidable to utilities.misc
* normalize_dimensions was added to assure that one is working with Points of
  matching dimensions
* docstring added to direction
* segment intersection works with differing dimensions; test added for that and
  other similar routines
* remove use of set() to remove duplicates; use uniq instead which makes
  routines canonical in behavior since it preserves order
* private _size property added to give the nominal dimension of a point
* are_coplanar added to Point; docstring moved there
* are_coplanar added to Point2D to return True for 2-D points
    maybe bad because then the general mutation of dimensions doesn't
    apply
* signature for is_concyclic changed to reflect that they are methods, not
  staticmethods: self added as first arg. raise error in Point's method if the
  dimension is not * supported to avoid recursion error.
* added docstring to orthogonal_direction
* unit method added to Point to normalize the coordinates of a point so it is in
  the same direction but a distance of 1 from the origin
* correct logic of Segment.perpendicular_line to not return None

smichr commented:
OK, I will check back next week. All significant lines that I added are covered with tests. The covereage of geometry (in general) is not that good, however.

siefkenj commented on sympy/geometry/line.py:
```diff
@@ -133,7 +128,35 @@ def p2(self):
+
+        sympy.geometry.point.Point.unit
+
+        """
```
Great!

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+                raise TypeError('Coordinates must be valid SymPy expressions.')
+
+        # adjust size
+        if len(coords) > dimension:
```
I'm okay with bumping up the dim, but I don't like the idea of bumping it down.  A user could always do `Point(a[:3])` if they wanted to reduce the dimension.

smichr replied:
It's only bumped down if the user says they want it so by supplying the `dim` keyword. We could add a `warn` keyword to allow a warning when dimensions are being reduced.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
     is_Point = True
 
+    @property
+    def _size(p):
```
If we remove the ability to reduce dimension, we shouldn't need this.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+            raise ValueError('Must provide 1 or more Points.')
+        if not all(isinstance(i, Point) for i in points):
+            raise TypeError('all elements must be Points')
+        minimal = not kwargs.get('max', False)
```
I don't like this idea :-/

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+        if not all(isinstance(i, Point) for i in points):
+            raise TypeError('all elements must be Points')
+        minimal = not kwargs.get('max', False)
+        dim = kwargs.get('dim',
```
I think `dim` should default to the dimension of the first argument.  To preserve compatibility, we should bump up the dim with a warning if it needs it.  If a `pad=True` argument is is given, then we bump up the dim without giving a warning.  We never bump down the dim.

I think silently changing the dimension of objects will lead to unexpected behavior from the user.  Unless we wanted to make the geometry module assume everything is in R^\infty and only print the non-zero entries...

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+
+        """
+        if not points:
+            return False
```
This should be `True`.  I don't think we need or want it, but we could have a `uniquely_coplanar` function, where `False` would be the appropriate return value.

smichr replied:
In order to answer the query of whether something lies in the same plane, one must have a plane to talk about...which doesn't exist with less than 3 pts. What about a ValueError in such cases?

siefkenj replied:
If I ask the question "Is the set of points X coplanar?" that logically translates to "Does there exist a plane P such that for every x in X, x is in P?"  If X is empty, this is trivially satisfied by any plane.  Of course, if we insist the plane is unique, it's no longer satisfied.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+            return False
+        points = [Point(i) for i in points]
+        dim = points[0].ambient_dimension
+        if not all(i.ambient_dimension == dim for i in points):
```
This would be a nice use of a `pad=False` kwarg.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+        if points[0].ambient_dimension < 4:
+            cls = points[0]
+            return cls.are_coplanar(*points)
+        raise NotImplementedError('dimension higher than 3')
```
We can do this for all dimensions with `affine_rank == 2`.  Is there a reason we wouldn't want that?

smichr replied:
I changed it to use that method.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -80,49 +82,225 @@ class Point(GeometryEntity):
+            return False
+        if args[0].ambient_dimension != 2:
+            raise NotImplementedError
         return args[0].is_concyclic(*args[1:])
```
Since this is explicitly 2D, I think we should do `Poin2D.is_concyclic` directly.

smichr replied:
One could have a circle rotated in space (in which case it becomes 3 dimensional).

smichr replied:
(I'm not sure we have the functionality to rotate 3D points into the xy plane so I just raise a NotImplementedError for those cases.)

siefkenj replied:
In this case, `Point.is_concyclic` should raise a `NotImplimentedError` and `Point2D.is_concyclic` should overload that function.  In any case, we might as well make this fully general by adding a `is_cospherical` that works in n dimensions and then `is_concyclic` would be `is_cospherical & is_coplanar`.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -155,32 +333,34 @@ def is_collinear(*args):
-        if len(args) == 0:
+        args = list(uniq(args))
+        if len(args) < 2:
             return False
```
I think this was discussed in a bug, but I cannot find it now.  I don't think this is the correct behavior.  Maybe I should open another bug report...Not sure the appropriate way to deprecate such a thing though...

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -233,21 +413,35 @@ def ambient_dimension(self):
     def orthogonal_direction(self):
-        """Returns a non-zero point in a direction
-        orthogonal to `self`."""
+        """Returns a non-zero point in the x-y plane that is
```
Methods that are 2d specific should go in `Point2D`.

smichr replied:
You could have a line bewteen two 3D points. I'm not sure what a line between the origin and a 4D point looks like but consider:

``` python
>>> p=Point(1,2,3,4)
>>> o=p.orthogonal_direction
>>> Line(Point(dim=4),p).angle_between(Line(Point(dim=4),o))
pi/2
```

(That's why I left it in the general Point class.)

siefkenj replied:
I guess I'm confused as to why you changed it from what it was.  It was a fully-general function.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -233,21 +413,35 @@ def ambient_dimension(self):
+    def unit(self):
+        """Return the Point that is in the same direction as `self`
+        and a distance of 1 from the origin"""
+        return self/self.distance(self.origin)
```
`self / abs(self)` is another way to say the same thing.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -622,28 +807,10 @@ class Point2D(Point):
+        if not kwargs.pop('_ok', False):
+            kwargs['dim'] = 2
+            args = Point(*args, **kwargs)
+        return GeometryEntity.__new__(cls, *args)
```
I like this much better.

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -687,7 +854,14 @@ def bounds(self):
+        points = [Point(i) for i in points]
+        if all(i.ambient_dimension == 2 for i in points):
+            return True  # all 2D points lie in the x-y plane
+        return Point.are_coplanar(*points)  # raise error there
```
I don't think this is right.  Since this is a method of `Point2D`, it should raise an error if any of `points` are not `Point2D`.  This may not raise an error...

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -905,25 +1075,10 @@ class Point3D(Point):
+        if not kwargs.pop('_ok', False):
+            kwargs['dim'] = 3
+            args = Point(*args, **kwargs)
+        return GeometryEntity.__new__(cls, *args)
```
Yay, again!

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -1062,52 +1217,24 @@ def are_collinear(*points):
-            except ValueError:
-                pass
-        raise ValueError('At least 3 non-collinear points needed to define plane.')
+            return False
```
See my prior disagreement about this.

smichr replied:
See the updated changes. I now answer the question via `affine_rank` which is going to reject cases where there are not enough points to define the object about which a query is being made.

siefkenj commented on sympy/geometry/tests/test_line.py:
```diff
@@ -403,17 +412,17 @@ def test_line3d():
+        Segment3D((3, 0), (4, 0))) == [Point((3, 0))]
     assert Segment3D((0, 0), (3, 0)).intersection(
-        Segment3D((2, 0), (5, 0))) == [Segment3D((3, 0), (2, 0))]
+        Segment3D((2, 0), (5, 0))) == [Segment2D((3, 0), (2, 0))]
```
I'm uncomfortable with downcasting.  I'll open a bug to discuss this.

siefkenj commented on sympy/geometry/ellipse.py:
```diff
@@ -845,13 +845,13 @@ def normal_lines(self, p, prec=None):
-        [Line2D(Point2D(-0.81, -2.7), Point2D(9/47, -21/17)),
-        Line2D(Point2D(1.5, -2.0), Point2D(32/13, -8/3))]
+        [Line2D(Point2D(-0.81, -2.7), Point2D(0.19, -1.2)),
+        Line2D(Point2D(1.5, -2.0), Point2D(2.5, -2.7))]
```
This is nice :-)

smichr replied:
I was happy to see that change come about, too.

siefkenj reviewed (commented):
I've reviewed `point.py` again.  Lot's of things to discuss.  Getting rid of `pointify` seems to have made things more straightforward.

smichr committed:
review modifications

siefkenj commented on sympy/geometry/point.py:
```diff
@@ -111,6 +112,8 @@ def __new__(cls, *args, **kwargs):
             if any(i for i in coords[dimension:]):
                 raise ValueError('too many nonzero dimensions')
+            if kwargs.get('warn', False):
+                raise ValueError('reducing dimensions')
```
I think this should be a `warnings.warn` instead of an error.  Though the `warnings` module isn't often used in sympy, I think it should be :-)

siefkenj commented:
I'll merge this and do some more hacking.

smichr commented:
> I'll merge this and do some more hacking.

sounds good...I'll try to keep current with the modifications.

siefkenj commented:
Thinking about this more, we cannot downstep the dimension or else something like

```
Line3D(Point(0,0,0), Point(1,1,1)).intersection(Line3D(Point(0,0,0), Point(1,1,2)))[0].z
```

would error since the `z` coordinate would have been eliminated.
